Add Rails/SafeNavigationWithBlank cop

gyfis opened the pull request:
This PR adds a new `Rails/SafeNavigatinoWithBlank` cop. 

It's a clone of the base Ruby cop [Lint/SafeNavigationWithEmpty](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/SafeNavigationWithEmpty) updated for `.blank?` in Rails, which has the same issue.

It's up for discussion if the cop should cover all usages of safe operator `&.` and `.blank?`, and not be limited only to conditionals. This could also be an option of the cop.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] ~Commit message starts with `[Fix #issue-number]` (if the related issue exists).~
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop-rails/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop-rails/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

koic commented on config/default.yml:
```diff
@@ -405,6 +405,11 @@ Rails/SafeNavigation:
 
+Rails/SafeNavigationWithBlank:
+  Description: 'Avoid `foo&.blank?` in conditionals.'
+  Enabled: true
```
I think this needs to be marked unsafe. And it may be useful to add the following example as a comment.

```diff
 Enabled: true
+# While the safe navigation operator is generally a good idea, when
+# checking `foo&.blank?` in a conditional, `foo` being `nil` will actually
+# do the opposite of what the author intends.
+#
+# foo&.blank? #=> nil
+# foo.blank? #=> true
+SafeAutoCorrect: false
 VersionAdded: '2.4'
```

koic replied:
Ah, I forgot to comment 💦 I think this cop can have autocorrect implemented.

gyfis replied:
Implemented in 5ee1d23f02b74cdffea0d7fe1eae866eb7d0fc4b.

koic commented on spec/rubocop/cop/rails/safe_navigation_with_blank_spec.rb:
```diff
@@ -0,0 +1,30 @@
+RSpec.describe RuboCop::Cop::Rails::SafeNavigationWithBlank do
+  subject(:cop) { described_class.new }
+
+  context 'target_ruby_version >= 2.3', :ruby23 do
```
Please remove the context because RuboCop Rails requires Ruby 2.3 or higher.
https://github.com/rubocop-hq/rubocop-rails/blob/v2.3.2/rubocop-rails.gemspec#L10

koic commented on lib/rubocop/cop/rails/safe_nagivation_with_blank.rb:
```diff
@@ -0,0 +1,41 @@
+      #
+      # @example
+      #   # bad
+      #   return if foo&.blank?
```
I think that `do_something if foo&.blank?` may be preferable to `return if foo&.blank?`. Because it is not specialized for `return`.

koic commented on lib/rubocop/cop/rails/safe_nagivation_with_blank.rb:
```diff
@@ -0,0 +1,41 @@
+      #   return unless foo.blank?
+      #
+      class SafeNavigationWithBlank < Cop
+        include RangeHelp
```
I think `include RangeHelp` can be removed.

gyfis commented:
@koic Thanks a lot for the review! 

I fixed the obvious remarks and also took a shot at implementing autocorrect for the cop - during that, I also fixed the pattern matcher a bit (I think it was wrong before). 

Let me know if I can do more on this PR!

koic commented on spec/rubocop/cop/rails/safe_navigation_with_blank_spec.rb:
```diff
@@ -0,0 +1,73 @@
+      RUBY
+    end
+
+    describe '#autocorrect' do
```
Thanks for the updating! Can you use `expect_correction` method for testing of autocorrect?

```diff
diff --git a/spec/rubocop/cop/rails/safe_navigation_with_blank_spec.rb b/spec/rubocop/cop/rails/safe_navigation_with_blank_spec.rb
index f334e9084..ec992c87d 100644
--- a/spec/rubocop/cop/rails/safe_navigation_with_blank_spec.rb
+++ b/spec/rubocop/cop/rails/safe_navigation_with_blank_spec.rb
@@ -6,15 +6,23 @@ RSpec.describe RuboCop::Cop::Rails::SafeNavigationWithBlank do
   context 'in a conditional' do
     it 'registers an offense on a single conditional' do
       expect_offense(<<~RUBY)
-        return unless foo&.blank?
-        ^^^^^^^^^^^^^^^^^^^^^^^^^ Avoid calling `blank?` with the safe navigation operator in conditionals.
+        do_something if foo&.blank?
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ Avoid calling `blank?` with the safe navigation operator in conditionals.
+      RUBY
+
+      expect_correction(<<~RUBY)
+        do_something if foo.blank?
       RUBY
     end

     it 'registers an offense on chained conditionals' do
       expect_offense(<<~RUBY)
-        return unless foo&.bar&.blank?
-        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Avoid calling `blank?` with the safe navigation operator in conditionals.
+        do_something if foo&.bar&.blank?
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Avoid calling `blank?` with the safe navigation operator in conditionals.
+      RUBY
+
+      expect_correction(<<~RUBY)
+        do_something if foo&.bar.blank?
       RUBY
     end

@@ -23,44 +31,6 @@ RSpec.describe RuboCop::Cop::Rails::SafeNavigationWithBlank do
         return if foo.blank?
       RUBY
     end
-
-    describe '#autocorrect' do
-      context 'single conditional' do
-        let(:source) do
-          <<~RUBY
-            do_something if foo&.blank?
-          RUBY
-        end
-
-        let(:corrected_source) do
-          <<~RUBY
-            do_something if foo.blank?
-          RUBY
-        end
-
-        it 'autocorrects' do
-          expect(autocorrect_source(source)).to eq(corrected_source)
-        end
-      end
-
-      context 'chained conditionals' do
-        let(:source) do
-          <<~RUBY
-            do_something if foo&.bar&.blank?
-          RUBY
-        end
-
-        let(:corrected_source) do
-          <<~RUBY
-            do_something if foo&.bar.blank?
-          RUBY
-        end
-
-        it 'autocorrects' do
-          expect(autocorrect_source(source)).to eq(corrected_source)
-        end
-      end
-    end
   end

   context 'outside a conditional' do
```

koic replied:
And can you squash your commits into one?

gyfis replied:
@koic `expect_correction` is neat, thanks! Updated & squashed commits.

koic commented:
Thanks!
